src/pages/StocktakeEditPage.js

```diff
@@ -51,130 +52,159 @@ import { gotoStocktakeManagePage } from '../navigation/actions';
-  const {
+    pageObject,
     data,
     dataState,
     sortBy,
     isAscending,
-    columns,
     modalKey,
-    pageInfo,
-    pageObject,
-    currentStocktakeItem,
     modalValue,
     keyExtractor,
+    PageActions,
+    columns,
+    getPageInfoColumns,
   } = state;
 
   const { isFinalised, comment, program } = pageObject;
 
+  // Listen to the stocktake become the top of the stack or being finalised,
+  // as these events are side-effects. Refreshing makes the state consistent again.
+  const refreshCallback = () => dispatch(PageActions.refreshData());
+  useRecordListener(refreshCallback, pageObject, 'Stocktake');
+  useNavigationFocus(refreshCallback, navigation);
+
+  // If the Stocktake is outdated, force a reset of the stocktake on mount.
   useEffect(() => {
-    if (stocktake.itemsOutdated.length && !isFinalised) {
-      dispatch(openModal(STOCKTAKE_OUTDATED_ITEM));
-    }
+    if (stocktake.isOutdated) dispatch(PageActions.openModal(MODAL_KEYS.STOCKTAKE_OUTDATED_ITEM));
   }, []);
 
+  const onFilterData = value => dispatch(PageActions.filterData(value));
+  const onEditBatch = rowKey => PageActions.openModal(MODAL_KEYS.EDIT_STOCKTAKE_BATCH, rowKey);
+  const onEditReason = rowKey => PageActions.openModal(MODAL_KEYS.STOCKTAKE_REASON, rowKey);
+  const onEditComment = value => dispatch(PageActions.editComment(value, 'Stocktake'));
+  const onCloseModal = () => dispatch(PageActions.closeModal());
+  const onApplyReason = ({ item }) => dispatch(PageActions.applyReason(item));
+  const onConfirmBatchEdit = () => dispatch(PageActions.closeAndRefresh());
+
+  const onResetStocktake = () =>
+    runWithLoadingIndicator(() => dispatch(PageActions.resetStocktake()));
+  const onManageStocktake = () =>
+    reduxDispatch(gotoStocktakeManagePage({ stocktake, stocktakeName: stocktake.name }));
+
   const renderPageInfo = useCallback(
-    () => <PageInfo columns={pageInfo(pageObject, dispatch)} isEditingDisabled={isFinalised} />,
+    () => (
+      <PageInfo
+        columns={getPageInfoColumns(pageObject, dispatch, PageActions)}
+        isEditingDisabled={isFinalised}
+      />
+    ),
     [comment, isFinalised]
   );
 
   const getAction = useCallback((colKey, propName) => {
     switch (colKey) {
       case 'countedTotalQuantity':
-        return editCountedTotalQuantity;
-      case 'modalControl':
-        return rowKey => openModal(EDIT_STOCKTAKE_BATCH, rowKey);
+        return PageActions.editCountedQuantity;
+      case 'batch':
+        return onEditBatch;
       case 'remove':
-        if (propName === 'onCheckAction') return selectRow;
-        return deselectRow;
+        if (propName === 'onCheckAction') return PageActions.selectRow;
+        return PageActions.deselectRow;
       case 'mostUsedReasonTitle':
-        return rowKey => openModal(STOCKTAKE_REASON, rowKey);
+        return onEditReason;
       default:
         return null;
     }
   });
 
+  const getModalOnSelect = () => {
+    switch (modalKey) {
```

Comment:
I find cases are easy to read and don't have the overhead of creating an object every invocation. This can also bring up some nasty bugs with different closures and fields on objects being undefined etc. In general I don't mind them - also have the benefit of a default that can be defined, other than `undefined` 🤷‍♂